pandera/engines/engine.py

```diff
@@ -0,0 +1,157 @@
+import functools
+import inspect
+import warnings
+from abc import ABCMeta
+from dataclasses import dataclass
+from typing import Any, Callable, Dict, List, Type, Union, get_type_hints
+
+import typing_inspect
+
+from pandera.dtypes_ import DataType
+
+
+@dataclass
+class _DtypeRegistry:
+    dispatch: Callable[[Any], DataType]
+    equivalents: Dict[Any, Type[DataType]]
+
+
+class Engine(ABCMeta):
```

Comment:
Awesome, are you already targeting a DataFrame backend other than pandas or just abstracting the current implementation?

`DataType.coerce` is linked to a particular DataFrame implementation. I'm not against separating validation and type into separate engines but they would probably reference each other.

Reply:
> are you already targeting a DataFrame backend other than pandas or just abstracting the current implementation?

I'm abstracting out the current implementation, but I am also gonna follow-up quickly with validators for `modin`, `dask`, or `koalas` (haven't decided yet 😄) since those would involve minimal code differences.

> DataType.coerce is linked to a particular DataFrame implementation. I'm not against separating validation and type into separate engines but they would probably reference each other.

Cool, on second thought let's hold off on the rename... I'll play around with the `Engine` class and see if it fits with the design I have so far and if not we can rename to `TypeEngine` and the validation engine could reference it.